File: app/src/main/java/com/gee12/mytetroid/activities/LogsActivity.java
package com.gee12.mytetroid.activities;

import android.net.Uri;
import android.os.AsyncTask;
import android.os.Bundle;
import android.text.TextUtils;
import android.view.View;
import android.view.WindowManager;
import android.widget.LinearLayout;
import android.widget.TextView;
import android.widget.Toast;

import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.gee12.mytetroid.R;
import com.gee12.mytetroid.adapters.TextAdapter;
import com.gee12.mytetroid.data.SettingsManager;
import com.gee12.mytetroid.logs.ILogger;
import com.gee12.mytetroid.logs.LogManager;
import com.gee12.mytetroid.utils.FileUtils;

import java.io.IOException;
import java.util.List;
import java.util.Locale;

/**
 * Активность для просмотра логов.
 */
public class LogsActivity extends AppCompatActivity {

    public static final int LINES_IN_RECYCLER_VIEW_ITEM = 10;
    
    private RecyclerView mRecycleView;
    private LinearLayout mLayoutError;
    private TextAdapter mTextAdapter;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_logs);
        Toolbar toolbar = findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);

        /*// убираем перенос слов, замедляющий работу
        if (Build.VERSION.SDK_INT >= 23) {
            mTextViewLogs.setBreakStrategy(Layout.BREAK_STRATEGY_SIMPLE);
            mTextViewLogs.setHyphenationFrequency(Layout.HYPHENATION_FREQUENCY_NONE);
        }*/
        this.mRecycleView = findViewById(R.id.recycle_view);
        mRecycleView.setLayoutManager(new LinearLayoutManager(this));
        this.mTextAdapter = new TextAdapter();
        mRecycleView.setAdapter(mTextAdapter);
        this.mLayoutError = findViewById(R.id.layout_read_error);

        if (SettingsManager.isWriteLogToFile(this)) {
            // читаем лог-файл
            LogManager.log(this, getString(R.string.log_open_log_file), ILogger.Types.INFO);
            new FileReadTask().execute(LogManager.getFullFileName());
        } else {
            // выводим логи текущего сеанса запуска приложения
            showBufferLogs();
        }
    }

    private void showBufferLogs() {
        String curLogs = LogManager.getBufferString();
        if (!TextUtils.isEmpty(curLogs)) {
            mLayoutError.setVisibility(View.GONE);
            mRecycleView.setVisibility(View.VISIBLE);
//            mTextAdapter.setItem(curLogs);
            try {
                // разбиваем весь поток строк логов на блоки
                List<String> logsBlocks = FileUtils.splitToBlocks(curLogs, LINES_IN_RECYCLER_VIEW_ITEM);
                mTextAdapter.setItems(logsBlocks);
            } catch (IOException e) {
                LogManager.log(this, R.string.log_error_logs_reading_from_memory, ILogger.Types.ERROR);
                mTextAdapter.setItem(curLogs);
            }
            scrollToBottom();
        } else {
            LogManager.log(this, getString(R.string.log_logs_is_missing), ILogger.Types.WARNING, Toast.LENGTH_SHORT);
        }
    }

    /**
     * Пролистывание в конец.
     */
    private void scrollToBottom() {
        mRecycleView.postDelayed(() -> mRecycleView.scrollToPosition(mTextAdapter.getItemCount() - 1), 100);
    }

    private void setProgresVisible(boolean isVis) {
        findViewById(R.id.progress_bar).setVisibility((isVis) ? View.VISIBLE : View.GONE);
    }


    /**
     * Задание, в котором выполняется чтение лог-файла.
     */
    private class FileReadTask extends AsyncTask<String, Void, FileReadTask.FileReadResult> {

        class FileReadResult {
//            PrecomputedTextCompat text;
            List<String> data;
            String text;
            boolean res;
//            FileReadResult(PrecomputedTextCompat text, boolean res) {
            FileReadResult(List<String> data, String text, boolean res) {
                this.data = data;
                this.text = text;
                this.res = res;
            }
        }

        @Override
        protected void onPreExecute() {
            getWindow().setFlags(WindowManager.LayoutParams.FLAG_NOT_TOUCHABLE,
                    WindowManager.LayoutParams.FLAG_NOT_TOUCHABLE);
            setProgresVisible(true);

        }

        @Override
        protected FileReadResult doInBackground(String... strings) {
            String fullFileName = strings[0];
            List<String> data = null;
            String text = null;
            boolean res = false;
            try {
//                text = FileUtils.readTextFile(Uri.parse(fullFileName));
                data = FileUtils.readTextFile(Uri.parse(fullFileName), LINES_IN_RECYCLER_VIEW_ITEM);
                res = true;
            } catch (IOException ex) {
                // ошибка чтения
                text = ex.getLocalizedMessage();
            }
//            PrecomputedTextCompat.Params params = mTextViewLogs.getTextMetricsParamsCompat();
//            PrecomputedTextCompat precomputedText = PrecomputedTextCompat.create(text, params);
//            return new FileReadResult(precomputedText, res);
            return new FileReadResult(data, text, res);
        }

        @Override
        protected void onPostExecute(FileReadResult res) {
            getWindow().clearFlags(WindowManager.LayoutParams.FLAG_NOT_TOUCHABLE);
            setProgresVisible(false);
            if (res.res) {
//                mTextViewLogs.setText(res.text);
                mTextAdapter.setItems(res.data);
            } else {
                String mes = String.format(Locale.getDefault(), "%s%s\n\n%s",
                        getString(R.string.log_file_read_error), LogManager.getFullFileName(), res.text);
                LogManager.log(LogsActivity.this, mes, ILogger.Types.ERROR, Toast.LENGTH_LONG);
                ((TextView) findViewById(R.id.text_view_error)).setText(mes);
                mLayoutError.setVisibility(View.VISIBLE);
                mRecycleView.setVisibility(View.GONE);
                // выводим логи текущего сеанса запуска приложения
                (findViewById(R.id.button_show_cur_logs)).setOnClickListener(v -> {
                    showBufferLogs();
                });
            }
            scrollToBottom();
        }
    }

}

File: app/src/main/java/com/gee12/mytetroid/ui/activities/LogsActivity.kt
package com.gee12.mytetroid.ui.activities

import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.text.TextUtils
import android.view.MenuItem
import android.view.View
import android.view.WindowManager
import android.widget.LinearLayout
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import androidx.lifecycle.lifecycleScope
import androidx.recyclerview.widget.RecyclerView
import com.gee12.mytetroid.ui.adapters.TextAdapter
import com.gee12.mytetroid.viewmodels.LogsViewModel
import com.gee12.mytetroid.R
import com.gee12.mytetroid.common.utils.FileUtils
import com.gee12.mytetroid.logs.Message
import com.gee12.mytetroid.viewmodels.LogsViewModel.*
import com.gee12.mytetroid.ui.TetroidMessage
import kotlinx.coroutines.launch
import org.koin.android.ext.android.inject
import java.io.IOException

/**
 * Активность для просмотра логов.
 */
class LogsActivity : AppCompatActivity() {

    private lateinit var recycleView: RecyclerView
    private lateinit var layoutError: LinearLayout
    private lateinit var textAdapter: TextAdapter

    private val viewModel: LogsViewModel by inject()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_logs)
        val toolbar = findViewById<Toolbar>(R.id.toolbar)
        setSupportActionBar(toolbar)
        supportActionBar!!.setDisplayHomeAsUpEnabled(true)

        viewModel.initialize()

        /*// убираем перенос слов, замедляющий работу
        if (Build.VERSION.SDK_INT >= 23) {
            mTextViewLogs.setBreakStrategy(Layout.BREAK_STRATEGY_SIMPLE);
            mTextViewLogs.setHyphenationFrequency(Layout.HYPHENATION_FREQUENCY_NONE);
        }*/
        recycleView = findViewById(R.id.recycle_view)
        textAdapter = TextAdapter()
        recycleView.adapter = textAdapter
        layoutError = findViewById(R.id.layout_read_error)

        lifecycleScope.launch {
            viewModel.logsEventFlow.collect { event -> onEvent(event) }
        }
        lifecycleScope.launch {
            viewModel.messageEventFlow.collect { showMessage(it) }
        }

        viewModel.load()

        viewModel.logDebug(getString(R.string.log_activity_opened_mask, javaClass.simpleName))
    }

    private fun onEvent(event: LogsEvent) {
        when (event) {
            LogsEvent.Loading.InProcess -> {
                window.setFlags(WindowManager.LayoutParams.FLAG_NOT_TOUCHABLE, WindowManager.LayoutParams.FLAG_NOT_TOUCHABLE)
                setProgresVisible(true)
            }

            is LogsEvent.Loading.Success,
            is LogsEvent.Loading.Failed -> {
                window.clearFlags(WindowManager.LayoutParams.FLAG_NOT_TOUCHABLE)
                setProgresVisible(false)

                when (event) {
                    is LogsEvent.Loading.Success -> {
                        textAdapter.setItems(event.data)
                    }
                    is LogsEvent.Loading.Failed -> {
                        val mes = "%s%s\n\n%s".format(getString(R.string.log_file_read_error), viewModel.logger.fullFileName, event.text)
                        (findViewById<View>(R.id.text_view_error) as TextView).text = mes
                        layoutError.visibility = View.VISIBLE
                        recycleView.visibility = View.GONE

                        // выводим логи текущего сеанса запуска приложения
                        findViewById<View>(R.id.button_show_cur_logs).setOnClickListener {
                            showBufferLogs()
                        }
                    }
                    else -> {}
                }

                scrollToBottom()
            }

            LogsEvent.ShowBufferLogs -> {
                showBufferLogs()
            }
        }
    }

    private fun showBufferLogs() {
        val curLogs = viewModel.getLogsBufferString()
        if (!TextUtils.isEmpty(curLogs)) {
            layoutError.visibility = View.GONE
            recycleView.visibility = View.VISIBLE

//            textAdapter.setItem(curLogs)
            try {
                // разбиваем весь поток строк логов на блоки
                // TODO: в IO поток ?
                val logsBlocks: List<String> = FileUtils.splitToBlocks(curLogs, LogsViewModel.LINES_IN_RECYCLER_VIEW_ITEM)
                textAdapter.setItems(logsBlocks)
            } catch (e: IOException) {
                viewModel.logError(R.string.log_error_logs_reading_from_memory)
                textAdapter.setItem(curLogs)
            }
            scrollToBottom()
        } else {
            viewModel.logWarning(getString(R.string.log_logs_is_missing))
        }
    }

    /**
     * Пролистывание в конец.
     */
    private fun scrollToBottom() {
        recycleView.postDelayed({ recycleView.scrollToPosition(textAdapter.itemCount - 1) }, 100)
    }

    private fun setProgresVisible(isVisible: Boolean) {
        findViewById<View>(R.id.progress_bar).visibility = if (isVisible) View.VISIBLE else View.GONE
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            android.R.id.home -> {
                finish()
                return true
            }
        }
        return super.onOptionsItemSelected(item)
    }

    private fun showMessage(message: Message) {
        TetroidMessage.show(this, message)
    }

    companion object {
        fun start(context: Context) {
            val intent = Intent(context, LogsActivity::class.java)
            context.startActivity(intent)
        }
    }
}